fastapi_sqla/__init__.py

```diff
@@ -147,24 +149,98 @@ class Page(Collection, Generic[T]):
     meta: Meta
 
 
-def _query_count(session: Session, query: DbQuery) -> int:
-    """Default function used to count items returned by a query.
+DbQuery = Union[LegacyQuery, Select]
+QueryCount = Callable[[SqlaSession, DbQuery], int]
+QueryCountDependency = Callable[..., QueryCount]
+PaginateSignature = Callable[[DbQuery], Page[T]]
 
-    Default Query.count is slower than a manually written query could be: It runs the
-    query in a subquery, and count the number of elements returned:
 
-    See https://gist.github.com/hest/8798884
-    """
-    return query.count()
+def query_count(session: Session, query: DbQuery) -> int:
+    @singledispatch
+    def _query_count(query: DbQuery):  # pragma no cover
+        "Dispatch on registered functions based on `query` type."
+        raise NotImplementedError(
+            f"no _query_count registered for type {type(query)!r}"
+        )
 
+    @_query_count.register
+    def sqla13_query_count(query: LegacyQuery) -> int:
+        """Default legacy function used to count items returned by a query.
 
-PaginateSignature = Callable[[DbQuery], Page[T]]
+        Default Query.count is slower than a manually written query could be: It runs the
+        query in a subquery, and count the number of elements returned:
+
+        See https://gist.github.com/hest/8798884
+        """
+        return query.count()
+
+    @_query_count.register
+    def sqla14_query_count(query: Select) -> int:
+        return session.execute(select(func.count()).select_from(query)).scalar()
+
+    return _query_count(query)
+
+
+@singledispatch
+def paginate_query(
+    query: DbQuery,
+    session: Session,
+    total_items: int,
+    offset: int,
+    limit: int,
+) -> Page[T]:  # pragma no cover
+    "Dispatch on registered functions based on `query` type"
+    raise NotImplementedError(f"no paginate_query registered for type {type(query)!r}")
+
+
+@paginate_query.register
+def _paginate_legacy(
+    query: LegacyQuery,
+    session: Session,
+    total_items: int,
+    offset: int,
+    limit: int,
+) -> Page[T]:
+    total_pages = math.ceil(total_items / limit)
+    page_number = offset / limit + 1
+    return Page[T](
+        data=query.offset(offset).limit(limit).all(),
+        meta={
+            "offset": offset,
+            "total_items": total_items,
+            "total_pages": total_pages,
+            "page_number": page_number,
+        },
+    )
+
+
+@paginate_query.register
+def _paginate(
+    query: Select,
+    session: Session,
+    total_items: int,
+    offset: int,
+    limit: int,
+) -> Page[T]:
```

Comment:
doesn't the generic need to be passed somehow? I feel like pydantic validation might not happen otherwise (how will it know otherwise what `T` is meant to be when doing the class instantiation? 

Check out how we do it on ER: https://github.com/dialoguemd/coredata/commit/5b5303c7cdb04ecb9d3b5c738917f1a18c98ef79

Reply:
I guess it doesn't matter for it to stay generic since fastAPI will re-parse it based on output model, but that means validation might not occur (I'm in fact very uncertain of what behaviour would be in that case)

Reply:
1. IMO scope of paginate is to ... paginate, not to do any type validation; This is the role of fastapi & pydantic;
2. Passing a `response_class` to `paginate` is not acceptable IMO;
3. TIL er is not using out of the box fastapi-sqla 😢

Reply:
> 1. IMO scope of paginate is to ... paginate, not to do any type validation; This is the role of fastapi & pydantic;

Still, fastapi-sqla must allow mypy to detect static type issues

Reply:
IMO: out of scope for current PR.

Reply:
related https://docs.sqlalchemy.org/en/14/orm/extensions/mypy.html

Reply:
shall we move forward? @vicrep

Reply:
Yes, out of scope I agree. In that case we could also consider returning a dict instead of doing an extra pydantic conversion :)